cpp/src/arrow/memory_pool.cc

```diff
@@ -88,22 +88,43 @@ MemoryPool::~MemoryPool() {}
 
 int64_t MemoryPool::max_memory() const { return -1; }
 
-class DefaultMemoryPool : public MemoryPool {
+///////////////////////////////////////////////////////////////////////
+// Helper tracking memory statistics
+
+class MemoryPoolStats {
  public:
-  DefaultMemoryPool() : bytes_allocated_(0) { max_memory_ = 0; }
+  MemoryPoolStats() : bytes_allocated_(0), max_memory_(0) {}
+
+  int64_t max_memory() const { return max_memory_.load(); }
+
+  int64_t bytes_allocated() const { return bytes_allocated_.load(); }
+
+  inline void UpdateAllocatedBytes(int64_t diff) {
+    auto allocated = bytes_allocated_.fetch_add(diff) + diff;
+    DCHECK_GE(allocated, 0) << "allocation counter became negative";
+    // "maximum" allocated memory is ill-defined in multi-threaded code,
+    // so don't try to be too rigorous here
+    if (diff > 0 && allocated > max_memory_) {
```

Comment:
Could use CAS here http://www.cplusplus.com/reference/atomic/atomic_compare_exchange_weak/

Reply:
AFAIK, that only works for equality testing.

Reply:
Atomics make my head hurt a little bit but it may be possible by writing a loop https://stackoverflow.com/a/16190791. I don't know that it's worth stressing out too much about

Reply:
Yes, it's possible to write a loop, but if you want to avoid two threads thrashing each other you probably want to limit the number of loop iterations and yield after that. I don't think it's worth the complication (and the overhead).

Reply:
Agreed.